app/models/tweet_fetcher.rb

```diff
@@ -0,0 +1,78 @@
+# frozen_string_literal: true
+
+# == Schema Information
+#
+# Table name: tweet_fetchers
+#
+#  id          :bigint           not null, primary key
+#  backoff     :integer          default(1)
+#  complete    :boolean          default(FALSE)
+#  created_at  :datetime         not null
+#  updated_at  :datetime         not null
+#  data_set_id :bigint
+#  user_id     :string
+#
+# Indexes
+#
+#  index_tweet_fetchers_on_data_set_id  (data_set_id)
+#
+# Foreign Keys
+#
+#  fk_rails_...  (data_set_id => data_sets.id)
+#
+
+# TweetFetcher fetches tweets for a single user. One DataSet spawns many
+# TweetFetchers, one per user_id in the cohort. This allows us to execute
+# tweet fetching out-of-band, so that we can be sensitive to rate limits.
+# When all of a DataSet's TweetFetchers are complete, the DataSet can be
+# finalized.
+class TweetFetcher < ApplicationRecord
+  belongs_to :data_set
+
+  def ingest
+    store_data(fetch_tweets)
+    # Make sure to mark this complete _before_ calling finish_when_ready, as
+    # finish_when_ready will check to see if all the data_set's TweetFetchers
+    # are complete!
+    self.update_attributes(complete: true)
+    data_set.finish_when_ready
+  rescue Twitter::Error::Unauthorized
+    data_set.unauthorized << user_id
```

Comment:
Is there a reason we don't generalize this mechanism to capture and collect all errors coming back from the TweetFetcher? I think Twitter has other unrecoverable errors if the user is suspended or deleted.

FYI, I had to add a retry mechanism to the Python stuff because Twitter often has recoverable errors. That can wait, but we might want to keep in mind that it'll be needed at some point.

Reply:
Because I haven't yet encountered them, and am unwilling to catch general Exceptions ;) And am not necessarily sure what the correct handling behavior is for the errors I haven't encountered. But I will look at all of the subclasses of Twitter::Error and see what makes sense.

Retry won't be hard given that I'm about to add jobs anyway.